src/ima_emulator.rs

```diff
@@ -0,0 +1,163 @@
+const FF_HASH: &[u8; 20] = &[0xffu8; 20];
+
+#[derive(Error, Debug)]
+enum ImaEmulatorError {
+    #[error("Invalid envvar")]
+    VarError(#[from] std::env::VarError),
+    #[error("TPM error")]
+    TssEsapiError(#[from] tss_esapi::Error),
+    #[error("Decoding error")]
+    FromHexError(#[from] hex::FromHexError),
+    #[error("I/O error")]
+    IoError(#[from] std::io::Error),
+    #[error("{0}")]
+    Other(String),
+}
+
+type Result<T> = std::result::Result<T, ImaEmulatorError>;
+
+fn ml_extend(
+    context: &mut Context,
+    ml: &str,
+    mut position: usize,
+    search_hash: Option<&Digest>,
+) -> Result<usize> {
+    let f = File::open(ml)?;
+    let mut reader = BufReader::new(f);
+    for line in reader.by_ref().lines().skip(position) {
+        let line = line?;
+        if line.is_empty() {
+            continue;
+        }
+        let tokens: Vec<&str> = line.splitn(5, ' ').collect();
+        if tokens.len() < 5 {
+            error!("invalid measurement list file line: -{}-", line);
+        }
+        position += 1;
+
+        let path = tokens[4];
+        let template_hash = hex::decode(tokens[1])?;
+        let template_hash = if template_hash == START_HASH {
+            Digest::try_from(&FF_HASH[..])
+        } else {
+            Digest::try_from(template_hash)
+        }?;
+
+        match search_hash {
+            None => {
+                println!(
+                    "extending hash {} for {}",
+                    hex::encode(template_hash.value()),
+                    &path
+                );
+                let mut vals = DigestValues::new();
+                vals.set(HashingAlgorithm::Sha1, template_hash);
+                // TODO: Add support for other hash algorithms
+                context.execute_with_nullauth_session(|ctx| {
+                    ctx.pcr_extend(PcrHandle::Pcr10, vals)
+                })?;
+            }
+            Some(search_hash) => {
+                let mut hasher = openssl::sha::Sha1::new();
+                hasher.update(START_HASH);
+                hasher.update(&template_hash);
+                let running_hash: Vec<u8> = hasher.finish().into();
+                let running_hash = Digest::try_from(running_hash)?;
+                let mut vals = DigestValues::new();
+                vals.set(HashingAlgorithm::Sha1, running_hash.clone());
+
+                if running_hash == *search_hash {
+                    println!("Located last IMA file updated: {}", path);
+                    return Ok(position);
+                }
+            }
+        }
+    }
+
+    if search_hash.is_some() {
+        return Err(ImaEmulatorError::Other(
+            "Unable to find current measurement list position, Resetting the TPM emulator may be neccesary".to_string()));
+    }
+
+    Ok(position)
+}
+
+fn main() -> std::result::Result<(), ImaEmulatorError> {
+    let tcti = Tcti::from_environment_variable()?;
+    let mut context = unsafe { Context::new(tcti) }?;
```

Comment:
we are generally trying to avoid unsafe (the compiler is currently complaining about this), is there not another approach we could leverage here?

Reply:
This will be unnecessary, once `tss-esapi` is upgraded to 6.x.